Sources/BuildSystemIntegration/ExternalBuildSystemAdapter.swift

```diff
@@ -281,3 +278,17 @@ actor ExternalBuildSystemAdapter {
     self.messagesToSourceKitLSPHandler.handle(OnBuildTargetDidChangeNotification(changes: nil))
   }
 }
+
+fileprivate extension URL {
+  /// If the path of this URL represents a directory, ensure that it has a trailing slash.
+  ///
+  /// This is important because if we form a file URL relative to eg. file:///tmp/a would assumes that `a` is a file
+  /// and use `/tmp` as the base, not `/tmp/a`.
+  var ensuringCorrectTrailingSlash: URL {
+    guard self.isFileURL else {
+      return self
+    }
+    // `URL(fileURLWithPath:)` checks the file system to decide whether a directory exists at the path.
+    return URL(fileURLWithPath: self.path)
```

Comment:
This will always `stat`. If we know a path is a directory, it is helpful to pass that as a hint to the initializer (`URL(filePath:isDirectoryHint:)`).

Reply:
We only call this when launching a BSP server, ie. once per workspace, so the extra `stat` doesn’t matter and I’d like to avoid issues caused by making wrong assumptions about whether we have a directory here.